model.py

```diff
@@ -0,0 +1,483 @@
+"""Model for the NFA ShotGrid Project Creator, written by Mervin van Brakel (2024)"""
+
+from __future__ import annotations
+
+import datetime
+import os
+import re
+from dataclasses import dataclass
+
+from shotgun_api3 import shotgun
+
+
+@dataclass
+class ProjectInformation:
+    """Dataclass used for storing all the information we need to create the project."""
+
+    username: str
+    project_name: str
+    has_production_code: bool
+    project_code: str
+    has_other_supervisors: bool
+    supervisor_list: list[dict]
+    render_engine: str
+    project_type: str
+    project_fps: int
+
+
+@dataclass
+class UserInformation:
+    """Dataclass used for storing information about the user who creates the project."""
+
+    sg_username: str
+    sg_user_id: str
+    student_year: int
+    student_graduation_year: int
+
+
+class ProjectCreatorModel:
+    """Model for the ShotGrid Project Creator.
+
+    This model handles all validation and ShotGrid inferfacing."""
+
+    def __init__(self) -> None:
+        """Initializes this class and stores the project dataclass."""
+        self.project_information = ProjectInformation(
+            "", "", True, "", False, [], "All", "Fiction", 25
+        )
+
+    def connect_to_shotgrid(self) -> (bool, str):
+        """Connects to ShotGrid database. Returns information based on
+        whether or not connection was successful.
+
+
+        Returns:
+            Bool: Whether or not connection was successful.
+            Str: Error message if connecting was not successful.
+        """
+        try:
+            self.shotgrid_connection = shotgun.Shotgun(
+                "https://nfa.shotgunstudio.com",
+                script_name="project_creation_V2",
+                api_key=os.environ["SHOTGRID_API_KEY"],
+            )
+
+            users = self.shotgrid_connection.find("HumanUser", [], ["name"])
+            self.usernames = [user["name"] for user in users]
+
+            projects = self.shotgrid_connection.find("Project", [], ["name"])
+            self.projects = [project["name"] for project in projects]
+
+            project_codes = self.shotgrid_connection.find(
+                "Project", [], ["sg_projectcode"]
+            )
+            self.project_codes = [
+                project_code["sg_projectcode"]
+                for project_code in project_codes
+            ]
+
+            return True, ""
+
+        except Exception as error:
+            return False, str(error)
```

Comment:
Returning tuples is not great, another approach that might be better is to raise the exception just when it happens. You can implement custom exceptions, then later where you need to show a message for example, you can do a try except and catch that exception.